src/bmp/BMPParser.cc

```diff
@@ -0,0 +1,403 @@
+#include "BMPParser.h"
+
+using namespace std;
+using namespace BMPParser;
+
+#define MAX_IMG_SIZE 10000
+
+#define E(cond, msg) if(cond) return setErr(msg)
+#define EU(cond, msg) if(cond) return setErrUnsupported(msg)
+#define EX(cond, msg) if(cond) return setErrUnknown(msg)
+
+#define GET_METHOD get
+
+#define I1() GET_METHOD<char>()
+#define U1() GET_METHOD<byte>()
+#define I2() GET_METHOD<int16_t>()
+#define U2() GET_METHOD<uint16_t>()
+#define I4() GET_METHOD<int32_t>()
+#define U4() GET_METHOD<uint32_t>()
+
+#define CALC_MASK(col) \
+  col##Mask = U4(); \
+  if(col##Mask == 0xffu) col##Mask = 0; \
+  else if(col##Mask == 0xff00u) col##Mask = 8; \
+  else if(col##Mask == 0xff0000u) col##Mask = 16; \
+  else if(col##Mask == 0xff000000u) col##Mask = 24; \
+  else EU(1, #col " mask");
+
+#define CHECK_OVERRUN(size, type) \
+  if(ptr + (size) - data > len){ \
+    setErr("unexpected end of file"); \
+    return type(); \
+  }
+
+Parser::Parser(){
+  status = Status::EMPTY;
+
+  w = 0;
+  h = 0;
+
+  data = nullptr;
+  ptr = nullptr;
+  imgd = nullptr;
+
+  op = "";
+  err = "";

```

Comment:
Up to you, but this entire ctor body could be done with default member initializers (i.e. in the header), and then you can do `Parser() = default`. Would save some lines.

```cpp
class Parser{
public:
  Parser() = default; 
  // ...
private:
  Status status = Status::EMPTY;
  // ...
}
```